builtin/credential/aws-iam/path_login.go

```diff
@@ -0,0 +1,376 @@
+package awsiam
+
+import (
+	"encoding/base64"
+	"encoding/xml"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+	"time"
+
+	//"github.com/hashicorp/vault/helper/policyutil"
+	"github.com/hashicorp/vault/helper/jsonutil"
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+)
+
+func pathLogin(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: "login$",
+		Fields: map[string]*framework.FieldSchema{
+			"role": &framework.FieldSchema{
+				Type: framework.TypeString,
+				Description: `Name of the role against which the login is being attempted.
+If 'role' is not specified, then the login endpoint looks for a role
+bearing the name of the IAM principal that is trying to login in.
+If a matching role is not found, the login fails.`,
+			},
+			"method": &framework.FieldSchema{
+				Type: framework.TypeString,
+				Description: `HTTP method to use for the AWS request. This must match what
+has been signed in the presigned request. Currently, POST is the only supported value`,
+			},
+			"url": &framework.FieldSchema{
+				Type: framework.TypeString,
+				Description: `Full URL against which to make the AWS request. If using a POST
+request with the action specified in the body, this should just be
+"/".`,
+			},
+			"body": &framework.FieldSchema{
+				Type: framework.TypeString,
+				Description: `Base64-encoded request body. This must match the request
+body included in the signature.`,
+			},
+			"headers": &framework.FieldSchema{
+				Type: framework.TypeString,
+				Description: `Base64-encoded JSON representation of the request headers.
+This must at a minimum include the headers over which AWS has included a
+signature.`,
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.UpdateOperation: b.pathLoginUpdate,
+		},
+
+		HelpSynopsis:    pathLoginSyn,
+		HelpDescription: pathLoginDesc,
+	}
+}
+
+func (b *backend) pathLoginUpdate(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+
+	// BEGIN boring data parsing
+	method := data.Get("method").(string)
+	if method == "" {
+		return logical.ErrorResponse("missing method"), nil
+	}
+
+	// In the future, might consider supporting GET
+	if method != "POST" {
+		return logical.ErrorResponse("Invalid method; currently only 'POST' is supported"), nil
+	}
+
+	rawUrl := data.Get("url").(string)
+	if rawUrl == "" {
+		return logical.ErrorResponse("missing url"), nil
+	}
+	parsedUrl, err := url.Parse(rawUrl)
+	if err != nil {
+		return logical.ErrorResponse("error parsing url"), nil
+	}
+
+	// TODO: There are two potentially valid cases we're not yet supporting that would
+	// necessitate this check being changed. First, if we support GET requests.
+	// Second if we support presigned POST requests
```

Comment:
Cool, though in thinking more, I think I should to update the docs to be explicit about this choice of not supporting presigned GET or POST requests (leaving this comment as a reminder to myself to do the docs update).